### PR TITLE
[IOHK-469] Adapt tracer-transformers to be thread-safe.

### PR DESCRIPTION
**Pull request by ksaric**
---
name: PR template
about: https://github.com/input-output-hk/iohk-monitoring-framework/issues/469

---

description
-----------

- [ ] https://github.com/input-output-hk/iohk-monitoring-framework/issues/469

checklist
---------

- [ ] compiles (`cabal new-clean; cabal new-build`)
- [ ] tests run successfully (`cabal new-test`)
- [ ] documentation added and created (`cd docs; nix-shell --run make`)
- [ ] link to an issue
- [ ] link to an epic
- [ ] add estimate points
- [ ] add milestone (the same as the linked issue)

### CODE REVIEW

File: tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs [hunk 1]
```diff
@@ -68,41 +71,67 @@ data OutcomeFidelity a
...
-                    liftIO (writeIORef s $ Just z)
+    traceOutcomes
+        :: MVar (Maybe (IntermediateValue a))
+        -> OutcomeEnhancedTracer m a
+        -> Tracer m a
+    traceOutcomes maybeInterValue tr = Tracer $ \a -> do
+      classifedObservable <- classifyObservable a
+      case classifedObservable of
+        OutcomeOther    -> traceWith tr $ Left a
+        -- modifyMVar_ is safer here.
```
Comment (ksaric) [inline tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs]: Maybe just use `modifyMVar_` here?

**Comment by ksaric** [comment]
@CodiePP I tested it using the examples.
I can add a test to demonstrate somewhat-correct behavior? :smile:

File: tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs [hunk 2]
```diff
@@ -20,8 +20,15 @@ module Control.Tracer.Transformers.ObserveOutcome
     , mkOutcomeExtractor
     ) where
 
+
 import           Control.Monad.IO.Class (MonadIO (..))
-import           Data.IORef (newIORef,readIORef, writeIORef)
+
+-- We really need to use Cardano.Prelude here and gain access to more
```
Comment (CodiePP) [inline tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs]: do we need this comment here?

File: tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs [hunk 3]
```diff
@@ -68,41 +74,73 @@ data OutcomeFidelity a
...
+        outcome         -> liftedModifyMVar_ maybeInterValue $ \observedResult -> case observedResult of
+                Nothing   -> outcomeWithoutValue outcome a
+                (Just b)  -> outcomeWithValue outcome a b
+      pure ()
+       where
+        -- If we don't have any intermediate values and the outcome is
+        -- @OutcomeStarts@, then we set the initial value inside the MVar.
+        outcomeWithoutValue :: OutcomeProgressionStatus -> a -> m (Maybe (IntermediateValue a))
+        outcomeWithoutValue OutcomeStarts a = do
+            -- Forces evaluation?!
```
Comment (CodiePP) [inline tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs]: yes. do we need this comment here?

File: tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs [hunk 4]
```diff
@@ -68,41 +74,73 @@ data OutcomeFidelity a
...
+            traceWith tr $ Right EndsBeforeStarted
+            -- We remove what we had been measuring, this should
+            -- probably be an error as well.
+            return $ Nothing
+
+        -- If we do have any intermediate values and the outcome is
+        -- @OutcomeStarts@, then we set the initial value inside the MVar.
+        outcomeWithValue :: OutcomeProgressionStatus -> a -> IntermediateValue a -> m (Maybe (IntermediateValue a))
+        outcomeWithValue OutcomeEnds a b = do
+            -- Forces evaluation?!
```
Comment (CodiePP) [inline tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs]: yes. do we need this comment here?

File: tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs [hunk 5]
```diff
@@ -68,41 +74,73 @@ data OutcomeFidelity a
...
+            !z <- captureObservableValue a
+            traceWith tr $ Left a
+            v <- computeOutcomeMetric a b z
+            traceWith tr $ Right (ProgressedNormally v)
+            return $ Nothing
+        outcomeWithValue _otherwise a b = do -- OutcomeStarts, this could be ignored since it "resets".
+            -- Forces evaluation?!
```
Comment (CodiePP) [inline tracer-transformers/src/Control/Tracer/Transformers/ObserveOutcome.lhs]: yes. do we need this comment here?

**Pull request merged**